databricks_cli/tunnel/server.py

```diff
@@ -0,0 +1,164 @@
+import logging
+import asyncio
+
+from urllib.parse import urljoin
+
+import aiohttp
+import socketio
+
+from aiohttp import BasicAuth
+from socketio.exceptions import BadNamespaceError
+from tornado.iostream import IOStream, StreamClosedError
+from tornado.tcpserver import TCPServer
+
+
+DBX_TUNNEL_SERVER_PORT = 6006
+
+
+logger = logging.getLogger(__name__)
+
+
+class TunnelConfig:
+    def __init__(self, databricks_config, org_id, cluster_id):
+        self.host = databricks_config.host
+        self.token = databricks_config.token
+        self.org_id = org_id
+        self.cluster_id = cluster_id
+        self.remote_port = DBX_TUNNEL_SERVER_PORT
+        # TODO(tunneling-cli): this assumes a multitenant workspace, we will need to confirm if
+        #  this is still the case for st workspace
+        self.endpoint_path = \
+            f"driver-proxy-api/o/{self.org_id}/{self.cluster_id}/{self.remote_port}/"
+
+    @property
+    def endpoint_url(self):
+        return urljoin(self.host, self.endpoint_path)
+
+    @property
+    def healthcheck_url(self):
+        return f"{self.endpoint_url}health"
+
+    @property
+    def socket_path(self):
+        return f"{self.endpoint_path}socket.io"
+
+
+class TunnelClient:
+    def __init__(self, stream: IOStream, session: aiohttp.ClientSession,
+                 tunnel_config: TunnelConfig, debug=False):
+        self.stream = stream
+        self.tunnel_config = tunnel_config
+        self.io = socketio.AsyncClient(http_session=session, logger=debug, engineio_logger=debug)
+        self.connection_established = False
+
+        @self.io.event
+        def connect():  # pylint: disable=unused-variable
+            self.connection_established = True
+            logger.info("Connection to socket.io server established with id %s", self.io.sid)
+
+        @self.io.on("response")
+        def response(msg):  # pylint: disable=unused-variable
+            data = msg.get("data")
+            # receive responses from tunnel server then write to tcp's stream
+            if isinstance(data, bytes):
+                self.stream.write(data)
+            else:
+                logger.warning("Received string message from byte-based stream: %s", data)
+
+    async def connect(self):
+        while not self.io.connected:
+            logger.info("initializing socket.io connection")
+            try:
+                headers = {
+                    'Content-Type': 'application/json'
+                }
+                await self.io.connect(self.tunnel_config.endpoint_url,
+                                      auth=None,
+                                      transports=["websocket"],
+                                      socketio_path=self.tunnel_config.socket_path,
+                                      headers=headers)
+            except socketio.exceptions.ConnectionError as e:
+                logger.error("Connection exception:")
+                logger.error(e)
+                raise e
```

Comment:
nit: You can just use `logger.exception` here. See https://www.loggly.com/blog/exceptional-logging-of-exceptions-in-python/

Reply:
I think you can log-and-raise here if the intention is to break the loop